File: Online-judge/leetcode/k-radius-subarray-averages.cpp

# include <bits/stdc++.h>
#define ll long long
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/k-radius-subarray-averages/description/
 * Running complexity O(n)
 * Memory complexity O(n)
*/

class Solution {
public:
    vector<int> getAverages(vector<int>& nums, int k) {
        int n = (int) nums.size();
        if (k == 0) {
            return nums;
        }

        vector<int> averages(n, -1);
        if (n < (2 * k +1)) {
            return averages;
        }

        vector<ll> prefix(n + 1, 0);
        for (int index = 0; index < n; index++) {
            prefix[index + 1] = prefix[index] + nums[index];  
        }

        for (int index = k; index < (n - k); index++) {
            int leftIndex = index - k, rightIndex = index + k;
            ll subArraySum = prefix[rightIndex + 1] - prefix[leftIndex];
            int average = subArraySum / (2 * k + 1);
            averages[index] = average;
        }
        return averages;
    }
};

File: Online-judge/leetcode/k-radius-subarray-averages.py
from typing import List

'''
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/k-radius-subarray-averages/description/
 * Time complexity: O(n).
 * Space complexity: O(n)
'''



class Solution:
    def getAverages(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        if k == 0:
            return nums
        averages = [-1]* n

        if n < (2 * k + 1):
            return averages

        windowSum = sum(nums[: 2 * k + 1])
        averages[k] = windowSum // (2 * k + 1)
        
        for i in range(2 * k + 1, n):
            windowSum = windowSum - nums[i - (2 * k + 1)] + nums[i]
            averages[i - k] = windowSum // (2 * k + 1)
        return averages